File: packages/sdk-typescript/src/resources/calls/calls.ts
import { Page, PagePromise } from '../../core/pagination';
import type { RequestExecutor, RequestOptions } from '../../core/types';
import type { Call, CallDetail, ListCallsQuery, MakeCallBody, MakeCallResult } from './types';

export class Calls {
  constructor(private readonly client: RequestExecutor) {}

  list(agentId: string, query: ListCallsQuery = {}): PagePromise<Call> {
    return new PagePromise(() => this.fetchPage(agentId, query.page ?? 1, query));
  }

  async make(
    agentId: string,
    body: MakeCallBody,
    options: RequestOptions = {},
  ): Promise<MakeCallResult> {
    const res = await this.client.request<MakeCallResult>({
      method: 'POST',
      path: `/agents/${agentId}/calls`,
      body,
      ...options,
    });
    return res.data;
  }

  async retrieve(agentId: string, callId: string): Promise<CallDetail> {
    const res = await this.client.request<CallDetail>({
      method: 'GET',
      path: `/agents/${agentId}/calls/${callId}`,
    });
    return res.data;
  }

  private async fetchPage(
    agentId: string,
    page: number,
    query: ListCallsQuery,
  ): Promise<Page<Call>> {
    const res = await this.client.request<Call[]>({
      method: 'GET',
      path: `/agents/${agentId}/calls`,
      query: {
        page,
        limit: query.limit,
        search: query.search,
        sort: query.sort,
        is_test_call: query.is_test_call,
        status: query.status,
      },
    });
    return new Page<Call>({
      data: res.data,
      meta: res.meta,
      requestId: res.requestId,
      fetcher: (next) => this.fetchPage(agentId, next, query),
    });
  }
}

File: packages/sdk-python/src/bimpeai/resources/calls.py
from __future__ import annotations

from .._request import RequestOptions
from ..pagination import AsyncPage, Page
from ..types.calls import Call, CallDetail, CallStatus, MakeCallBody, MakeCallResult
from ._specs import (
    AsyncTransport,
    SyncTransport,
    list_calls_spec,
    make_call_spec,
    retrieve_call_spec,
)


class Calls:
    def __init__(self, client: SyncTransport) -> None:
        self._client = client

    def list(
        self,
        agent_id: str,
        *,
        page: int = 1,
        limit: int | None = None,
        search: str | None = None,
        sort: str | None = None,
        is_test_call: bool | None = None,
        status: CallStatus | None = None,
    ) -> Page[Call]:
        def fetch(current: int) -> Page[Call]:
            resp = self._client.request(
                list_calls_spec(
                    agent_id,
                    page=current,
                    limit=limit,
                    search=search,
                    sort=sort,
                    is_test_call=is_test_call,
                    status=status,
                )
            )
            data = [Call.model_validate(item) for item in resp.data]
            return Page(data=data, meta=resp.meta, request_id=resp.request_id, fetcher=fetch)

        return fetch(page)

    def make(
        self,
        agent_id: str,
        body: MakeCallBody,
        *,
        idempotency_key: str | None = None,
        timeout: float | None = None,
        max_retries: int | None = None,
        headers: dict[str, str] | None = None,
    ) -> MakeCallResult:
        options = RequestOptions(
            idempotency_key=idempotency_key,
            timeout=timeout,
            max_retries=max_retries,
            headers=headers,
        )
        resp = self._client.request(make_call_spec(agent_id, dict(body), options))
        return MakeCallResult.model_validate(resp.data)

    def retrieve(self, agent_id: str, call_id: str) -> CallDetail:
        resp = self._client.request(retrieve_call_spec(agent_id, call_id))
        return CallDetail.model_validate(resp.data)


class AsyncCalls:
    def __init__(self, client: AsyncTransport) -> None:
        self._client = client

    async def list(
        self,
        agent_id: str,
        *,
        page: int = 1,
        limit: int | None = None,
        search: str | None = None,
        sort: str | None = None,
        is_test_call: bool | None = None,
        status: CallStatus | None = None,
    ) -> AsyncPage[Call]:
        async def fetch(current: int) -> AsyncPage[Call]:
            resp = await self._client.request(
                list_calls_spec(
                    agent_id,
                    page=current,
                    limit=limit,
                    search=search,
                    sort=sort,
                    is_test_call=is_test_call,
                    status=status,
                )
            )
            data = [Call.model_validate(item) for item in resp.data]
            return AsyncPage(data=data, meta=resp.meta, request_id=resp.request_id, fetcher=fetch)

        return await fetch(page)

    async def make(
        self,
        agent_id: str,
        body: MakeCallBody,
        *,
        idempotency_key: str | None = None,
        timeout: float | None = None,
        max_retries: int | None = None,
        headers: dict[str, str] | None = None,
    ) -> MakeCallResult:
        options = RequestOptions(
            idempotency_key=idempotency_key,
            timeout=timeout,
            max_retries=max_retries,
            headers=headers,
        )
        resp = await self._client.request(make_call_spec(agent_id, dict(body), options))
        return MakeCallResult.model_validate(resp.data)

    async def retrieve(self, agent_id: str, call_id: str) -> CallDetail:
        resp = await self._client.request(retrieve_call_spec(agent_id, call_id))
        return CallDetail.model_validate(resp.data)
